src/primitive/circle.rs

```diff
@@ -0,0 +1,201 @@
+//! Circle primitive
+
+use cgmath::{BaseFloat, Point2, Vector2};
+use cgmath::prelude::*;
+
+use {Aabb2, Ray2};
+use prelude::*;
+
+/// Circle primitive
+#[derive(Debug, Clone, PartialEq)]
+#[cfg_attr(feature = "eders", derive(Serialize, Deserialize))]
+pub struct Circle<S> {
+    /// Radius of the circle
+    pub radius: S,
+}
+
+impl<S> Circle<S> {
+    /// Create a new circle primitive
+    pub fn new(radius: S) -> Self {
+        Self { radius }
+    }
+}
+
+impl<S> SupportFunction for Circle<S>
+where
+    S: BaseFloat,
+{
+    type Point = Point2<S>;
+
+    fn support_point<T>(&self, direction: &Vector2<S>, transform: &T) -> Point2<S>
+    where
+        T: Transform<Point2<S>>,
+    {
+        let direction = transform
+            .inverse_transform()
+            .unwrap()
+            .transform_vector(*direction);
+        transform.transform_point(Point2::from_vec(direction.normalize_to(self.radius)))
+    }
+}
+
+impl<S> HasAabb for Circle<S>
+where
+    S: BaseFloat,
+{
+    type Aabb = Aabb2<S>;
+
+    fn get_bound(&self) -> Aabb2<S> {
+        Aabb2::new(
+            Point2::from_value(-self.radius),
+            Point2::from_value(self.radius),
+        )
+    }
+}
+
+impl<S> Discrete<Ray2<S>> for Circle<S>
+where
+    S: BaseFloat,
+{
+    fn intersects(&self, r: &Ray2<S>) -> bool {
+        let s = self;
+        let l = Vector2::new(-r.origin.x, -r.origin.y);
+        let tca = l.dot(r.direction);
+        if tca < S::zero() {
+            return false;
+        }
+        let d2 = l.dot(l) - tca * tca;
+        if d2 > s.radius * s.radius {
+            return false;
+        }
+        return true;
```

Comment:
no need for `return` at the end, could just end up with `d2 <= s.radius * s.radius`